### PR TITLE
PWN-3456 - History buffer implementation

### PR DESCRIPTION
**Pull request by davran1001**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3456
https://p2pvalidator.atlassian.net/browse/PWN-3657
Implement history buffer implementation

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryPresenter.kt [hunk 1]
```diff
@@ -102,18 +102,18 @@ class TokenHistoryPresenter(
...
-            )
-            val fetchedItems = historyInteractor.loadTransactionHistory(
-                tokenPublicKey = token.publicKey,
-                signaturesWithStatus = signatures,
-                forceRefresh = isRefresh
-            )
-
-            lastTransactionSignature = fetchedItems.last().signature
-            transactions.addAll(fetchedItems)
+//            val signatures = historyInteractor.loadSignaturesForAddress(
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryPresenter.kt:105]: remove or ancomment?

File: app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt [hunk 2]
```diff
@@ -90,7 +136,7 @@ class HistoryInteractor(
                     swapTransaction.alternateDestination
                 )
             }
-
+        Timber.tag("TokenHistoryBuffer").d("AccountInfoSize = " + accountsInfoIds.size.toString())
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt]: Remove

File: app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt [hunk 3]
```diff
@@ -5,65 +5,111 @@ import org.p2p.solanaj.kits.transaction.TransactionDetails
...
+        return loadTransactions(signatures)
+    }
+
+    private suspend fun loadAllSignatures(): MutableList<RpcTransactionSignature> {
+        val transactionsSignatures = mutableListOf<RpcTransactionSignature>()
+        while (true) {
+            val firstItem = multipleStreamSource.currentItem() ?: break
+            val time = (firstItem.streamSource?.blockTime ?: -1) - (60 * 60 * 24)
+
+            while (true) {
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt:81]: Please add comments to the class, how the history works. 
And add todo about the refactoring

File: app/src/main/java/org/p2p/wallet/history/interactor/stream/MultipleStreamSource.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            fillBuffer(configuration)
+        }
+        if (buffer.isEmpty()) {
+            return null
+        }
+        return buffer.removeAt(0)
+    }
+
+    override fun reset() {
+    }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/interactor/stream/MultipleStreamSource.kt:49]: Why all reset functions are empty?

File: app/src/main/java/org/p2p/wallet/history/repository/remote/TransactionDetailsRpcRepository.kt [hunk 5]
```diff
@@ -20,6 +21,7 @@ class TransactionDetailsRpcRepository(
         userPublicKey: String,
         transactionSignatures: List<RpcTransactionSignature>
     ): List<TransactionDetails> {
+        Timber.tag("TokenHistoryBuffer").d("signatures size = ${transactionSignatures.size}")
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/TransactionDetailsRpcRepository.kt:24]: clean logs

File: app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryFragment.kt [hunk 6]
```diff
@@ -165,6 +165,13 @@ class TokenHistoryFragment :
 
     override fun showPagingState(newState: PagingState) {
         historyAdapter.setPagingState(newState)
+        with(binding) {
+            shimmerView.isVisible = newState == PagingState.InitialLoading
+            refreshLayout.isVisible = newState != PagingState.InitialLoading
+            emptyStateLayout.isVisible = newState == PagingState.Idle && historyAdapter.isEmpty()
+            historyRecyclerView.isVisible =
+                (newState == PagingState.Idle && !historyAdapter.isEmpty()) || newState == PagingState.Loading
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryFragment.kt:173]: Why do you make such complex conditions for simple states? 
You'd rather to check each state and update UI according to it. 
This one is hard to support.

The best way will be controlling the UI from presenter

File: app/src/main/res/layout/fragment_token_history.xml [hunk 7]
```diff
@@ -130,6 +132,125 @@
...
+
+                <include
+                    layout="@layout/item_history_shimmer"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginTop="16dp" />
+
+                <ImageView
+                    android:layout_width="104dp"
+                    android:layout_height="16dp"
```
Comment (1sakov) [inline app/src/main/res/layout/fragment_token_history.xml:190]: What is this? Weird size

File: app/build.gradle [hunk 8]
```diff
@@ -263,6 +263,8 @@ dependencies {
     //https://facebook.github.io/shimmer-android/
     implementation 'com.facebook.shimmer:shimmer:0.5.0'
 
+    implementation 'com.localebro:okhttpprofiler:1.0.8'
```
Comment (1sakov) [inline app/build.gradle:266]: Do we need that? Or you used it only for debug purposes?

**Review by 1sakov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryFragment.kt (hunk 6 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryFragment.kt:173]: can we refactor this later ?

File: app/src/main/java/org/p2p/wallet/infrastructure/network/interceptor/RpcInterceptor.kt [hunk 9]
```diff
@@ -42,15 +42,17 @@ open class RpcInterceptor(
         return if (response.isSuccessful) {
             handleResponse(response)
         } else {
-            throw extractGeneralException(response.body!!.string())
+            val json = getRequestJson(request)
+            Timber.tag("TokenHistoryBuffer").d(json.toString())
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/interceptor/RpcInterceptor.kt:46]: do we need this log?

File: app/src/main/res/layout/fragment_token_history.xml [hunk 10]
```diff
@@ -130,6 +133,117 @@
             android:layout_marginTop="16dp"
             android:background="@drawable/bg_buttons"
             app:layout_constraintTop_toBottomOf="@id/emptyStateLayout" />
+
+        <com.facebook.shimmer.ShimmerFrameLayout
```
Comment (gslevinkov) [inline app/src/main/res/layout/fragment_token_history.xml:137]: can you extract shimmering as <include> or <merge> layout

**Review by gslevinkov — APPROVED** [review]

**Review by 1sakov — APPROVED** [review]

**Review by kuznetsov — APPROVED** [review]

**Pull request merged**